kpet/cmd_tree.py

```diff
@@ -41,7 +45,18 @@ def main(args):
     database = data.Base(args.db)
     if args.action == 'list':
         regex = re.compile(args.regex) if args.regex else None
-        for tree in sorted(database.trees.keys()):
+        arch_regex = re.compile(args.arch) if args.arch else None
+        if arch_regex:
+            matching_trees = []
+            for tree in database.trees:
+                for arch in database.trees[tree]['arches']:
+                    if arch_regex.fullmatch(arch):
+                        matching_trees.append(tree)
+            matching_trees = list(dict.fromkeys(matching_trees))
+        else:
+            matching_trees = database.trees.keys()
+
+        for tree in sorted(matching_trees):
```

Comment:
With the above changes to regex initialization, you can replace this loop and the code above with something like this:
```python
tree_names=[]
for name, value in database.trees.items():
    if regex.fullmatch(name) and
       filter(arch_regex.fullmatch, value['arches']):
        tree_names.append(name)
for name in sorted(tree_names):
    print(name)
```

All of the above goes for `cmd_arch.py` too, more-or-less.